### ISSUE TITLE
Multiple Devs in One Location on map

### ISSUE THREAD

**Issue by daniel-chenery** [umbraco/OurUmbraco#207]
We've both set our location the the [office](https://www.google.co.uk/maps/place/btc,+Business+and+Technology+Centre/@51.9016363,-0.2181888,17z/data=!3m1!4b1!4m5!3m4!1s0x487631b625610a87:0x667a157c2580bd5a!8m2!3d51.901633!4d-0.2160001).

The Umbraco map correctly displays the number "2" for this location, however clicking on it zooms the map excessively and fails to show anybody in that location.
It works fine if you zoom out slightly :)

![image](https://user-images.githubusercontent.com/3447210/37972788-fa9c7b74-31d0-11e8-8b40-187af2ea2974.png)

**Comment by abjerner** [comment]
Just throwing this in here as a possible fix:

https://github.com/jawj/OverlappingMarkerSpiderfier

I researched it once for a project, but ended up with another solution, so I haven't used it myself. The demos appear to just use regular markers, so I'm also not sure how it will work with clusters.

**Comment by warrenbuckley** [comment]
Oh neat find @abjerner I will have to look & investigate that.
Think the first thing is to fix this bug/issue & then take a look at implementing this library if that is the solution. As London has quite a few that overlap not just 2 or so

**Comment by warrenbuckley** [comment]
@nul800sebastiaan the underlying issue is due to our rounding of the lat & lon it seems if I revert this on my local copy then this works fine.

Will see if I can find a solution & get this sorted whilst making sure the markers are rounded & not too precise from the API controller

**Comment by nul800sebastiaan** [comment]
Aha.. okay cool, maybe we should add in a random fudge factor after the 2nd digit then, location will become a random spot in the 1.1km area then.

**Comment by daniel-chenery** [comment]
Nothing like magic numbers to fix a problem 🤣

**Comment by nul800sebastiaan** [comment]
Random numbers! ;-)

I mostly want to make sure we're not disclosing people's locations too much down to the centimeter.

**Comment by abjerner** [comment]
But ... two sorta-random locations may still share the same coordinates - or be really close.

I may however have a solution that I've used in the past. I can try to look into that later ;)

**Comment by nul800sebastiaan** [comment]
You're right! I got it to work now with randomized locations but of course then this happens (and sometimes the markers are not even in the viewport any more.

![mapsclusterfck](https://user-images.githubusercontent.com/304656/38028214-84dde1a8-3292-11e8-84aa-9260dd5e2773.gif)

**Comment by abjerner** [comment]
Okay - so the solution I was talking works by grouping locations (or in this case members) with the exact same coordinates, and then run the latitude and longitude through the two methods below. It will do nothing for the first location (where `group` is zero), but the remaining locations will be placed around the original coordinates.

Here illustrated with my awesome Paint skills:

![image](https://user-images.githubusercontent.com/3634580/38030062-dbf93564-3297-11e8-95c6-7df8e1c380d6.png)

The methods probably needs a bit of adjustments - and technically the locations placed around the original coordinates may also overlap other locations in close proximity, so maybe we should initially group coordinates in close proximity. Again, there are a few things to take into consideration.

I'll try to play around with this (hopefully the Our test database will have sufficient data to test with).


```
private void AdjustLatitude(ref double latitude, int count) {

    const double factor = 0.000025;

    if (count >= 1 && count <= 2) latitude += factor;
    if (count >= 4 && count <= 6) latitude -= factor;

    if (count >= 9 && count <= 12) latitude += (2 * factor);
    if (count == 13 || count == 23) latitude += factor;

    if (count == 15 || count == 21) latitude -= factor;
    if (count >= 16 && count <= 20) latitude += (2 * factor);

}

private void AdjustLongitude(ref double longitude, int count) {

    const double factor = 0.000050;

    if (count >= 2 && count <= 4) longitude += factor;
    if (count >= 6 && count <= 8) longitude -= factor;

    if (count >= 12 && count <= 16) longitude += (2 * factor);
    if (count >= 20 && count <= 54) longitude -= (2 * factor);

    if (count == 11 || count == 17) longitude += factor;
    if (count == 9 || count == 19) longitude -= factor;

}
```

**Comment by abjerner** [comment]
Btw - is you GIF from London? I suddenly feel like going to **BEST KEBAB** 😄 

Also, what tool do you use for converting a screencast to a GIF?

**Comment by nul800sebastiaan** [comment]
Haha, yes! Send more Kebab. It's actually near where Warren lives, South-End.

Camtasia makes the gifs for me, but costs a bit of course.

**Comment by nul800sebastiaan** [comment]
btw I've just added a max zoom level to the whole map, so we could add this as well, to uncluster at the max zoom level:

https://github.com/googlemaps/v3-utility-library/blob/master/markerclusterer/src/markerclusterer.js#L338

Not sure how well your adjustments will work if just zooming in on the central point of "London" where there 166 people.

![image](https://user-images.githubusercontent.com/304656/38033679-d68d2d34-32a0-11e8-9074-01b6f5cee11e.png)

**Comment by warrenbuckley** [comment]
I would say the original issue is fixed that avatars under the map disappeared when you zoomed in super close @nul800sebastiaan & we should close this?

Perhaps create a new issue & link back here to talk about how to improve the map UI when there are 150+ users at same location, perhaps from @abjerner suggestion/s
